### PR TITLE
common.sh fix to detect ip instead of ifconfig

### PR DESCRIPTION
**Pull request by rphillips**
**What this PR does / why we need it**: The IP address detection recently [added](https://github.com/kubernetes/kubernetes/commit/ca76002) to build/common.sh to whitelist the host IP will error if ifconfig is not installed. This patch detects if `ip` is installed and then uses `ip` instead of `ifconfig`.

**Which issue this PR fixes**

**Special notes for your reviewer**:

**Release note**:

/cc @zhouhaibing089

### CODE REVIEW

File: build/common.sh [hunk 1]
```diff
@@ -317,6 +317,10 @@ function kube::build::has_docker() {
   which docker &> /dev/null
 }
 
+function kube::build::has_ip() {
```
Comment (cblecker) [inline build/common.sh]: This function name isn't descriptive (having an IP is different than having the ip tool), and lacks a comment similar to other functions in this script.

File: build/common.sh [hunk 2]
```diff
@@ -628,11 +632,15 @@ function kube::build::rsync_probe {
 # This will set the global var KUBE_RSYNC_ADDR to the effective port that the
 # rsync daemon can be reached out.
 function kube::build::start_rsyncd_container() {
+  IPTOOL=ifconfig
+  if kube::build::has_ip ; then
```
Comment (cblecker) [inline build/common.sh]: This prefers ip over ifconfig. Also, which "ip" tool are you referring to? The name is pretty generic. Does it accept the same syntax? I would want to try ifconfig first, and ifconfig doesn't work, then try another tool.

File: build/common.sh [hunk 3]
```diff
@@ -628,11 +632,15 @@ function kube::build::rsync_probe {
...
+    IPTOOL="ip address"
+  fi
   kube::build::stop_rsyncd_container
   V=3 kube::log::status "Starting rsyncd container"
   kube::build::run_build_command_ex \
     "${KUBE_RSYNC_CONTAINER_NAME}" -p 127.0.0.1:${KUBE_RSYNC_PORT}:${KUBE_CONTAINER_RSYNC_PORT} -d \
-    -e ALLOW_HOST="$(ifconfig | grep -Eo 'inet (addr:)?([0-9]*\.){3}[0-9]*' | grep -Eo '([0-9]*\.){3}[0-9]*' | grep -v '127.0.0.1')" \
+    -e ALLOW_HOST="$(${IPTOOL} | grep -Eo 'inet (addr:)?([0-9]*\.){3}[0-9]*' | grep -Eo '([0-9]*\.){3}[0-9]*' | grep -v '127.0.0.1')" \
```
Comment (cblecker) [inline build/common.sh]: Perhaps IP detection should be offloaded into it's own function (with local vars), and then called from here, rather than complicating this bit with another global environment variable.

**Review by cblecker — CHANGES_REQUESTED** [review]

File: build/common.sh [hunk 4]
```diff
@@ -317,6 +317,10 @@ function kube::build::has_docker() {
   which docker &> /dev/null
 }
 
+function kube::build::has_ip() {
+  which ip &> /dev/null
```
Comment (euank) [inline build/common.sh]: This matches the style of the other `has_...` functions, but technically `command -v ip` is preferable. Even `type ip` is better than `which` given `which` has some weird historical behavior.

I think if we want to change this, it should be a separate PR that changes all of them probably though.

**Comment by cblecker** [comment]
/lgtm cancel
/assign
This PR needs changes.

File: build/common.sh (hunk 3 above)
Comment (euank) [reply build/common.sh]: This pr makes the state of the world no worse IMO.

I think refactoring this should be a separate pr from fixing its reliance on a deprecated tool. ifconfig is deprecated at this point.

File: build/common.sh (hunk 2 above)
Comment (euank) [reply build/common.sh]: `ifconfig` comes from the `net-tools` package, which is in the process of being deprecated.

See this article https://lwn.net/Articles/710533/

the `iproute2` tools, which includes `ip`, is the way of the world these days.

File: build/common.sh (hunk 1 above)
Comment (rphillips) [reply build/common.sh]: `ip` is quite vague in the Linux world, but it is appropriate in this context based on the has_docker function

Comment (cblecker) [reply build/common.sh]: Perhaps "has_iproute2" which is the package name?

File: build/common.sh (hunk 3 above)
Comment (rphillips) [reply build/common.sh]: I think a separate PR makes sense to improve the function, but this PR is fixing a build regression introduced recently.

File: build/common.sh (hunk 4 above)
Comment (cblecker) [reply build/common.sh]: I'd suggest doing something like `ip -V | grep -o 'iproute2'`.. this not only ensures there's a binary named "ip", but it's actually the "ip" tool from the iproute2 package.

File: build/common.sh (hunk 2 above)
Comment (cblecker) [reply build/common.sh]: Got it. Yes, then I'd agree with this. However "ip" is a generic binary name.. I've suggested above to verify that the ip tool is actually the one from the iproute2 package, and if not, then we fail back to ifconfig.

File: build/common.sh (hunk 3 above)
Comment (cblecker) [reply build/common.sh]: It does make it worse if we blindly trust that a binary named "ip" is preferred over ifconfig. If we add extra validation that "ip" is actually the iproute2 tool, then I'm okay with this. Refactor would also be a good idea, but it can be a separate PR if we validate the binary.

File: build/common.sh (hunk 2 above)
Comment (euank) [reply build/common.sh]: I disagree that `ip` is a generic binary name. It's a standard linux thing. Any distro which has a different, non-compatible, command in its place is broken and should fix their setup.

Comment (cblecker) [reply build/common.sh]: Linux isn't the only operating system out there (we verify binaries in other parts for darwin compatibility), and you also have to contend with the fact that these are local developer environments. I personally have a utility already on my darwin system that is named "ip" that is not the iproute2 utility.

I believe it's worth verifying that the "ip" binary you are going to call is the intended binary, and I've suggested a method of verifying the binary here: https://github.com/kubernetes/kubernetes/pull/53500#discussion_r143051468

File: build/common.sh (hunk 4 above)
Comment (euank) [reply build/common.sh]: We don't do this elsewhere and I don't think we should do it here.

If someone wishes to have a roughly compatible `ip` tool with different `-V` output (such as [iproute2mac](https://github.com/brona/iproute2mac)), that's fine. Let them.

Second guessing users to that degree is the wrong solution to the problem.

The right solution is relying on users having a reasonable environment.

File: build/common.sh (hunk 2 above)
Comment (euank) [reply build/common.sh]: I don't think it's a good idea to try and second guess the user like that.
I would be okay with hiding it in an `if kube::build::is_osx` instead, though I could go either way since I know there are at least a couple OSX utilities which provide a sufficiently compatible `ip` command for `ip a`, which muddies the waters some.

I think you might be better to have your `ip` command be an alias that exists in your interactive shell, or if it was provided by a third party to let them know it's a bad name to use considering other shell scripts will make this same check.

File: build/common.sh (hunk 4 above)
Comment (cblecker) [reply build/common.sh]: We do do this elsewhere, for example, with sed: https://github.com/kubernetes/kubernetes/search?utf8=%E2%9C%93&q=gsed&type=

This is attempting to solve a build environment break, and it would be wrong to break other users environments just to fix yours. We should either default to ifconfig, or verify that the "ip" binary is the one we intend to call. Blindly defaulting to "ip" isn't the right choice in my opinion.

Comment (euank) [reply build/common.sh]: In the case of sed, it's feature-detection for a common machine configuration.

> It would be wrong to break other users environments just to fix yours.

Linux machines that have a tool named `ip` available to noninteractive shells which isn't *the* `ip` tool are already broken.
We have no obligation to write overtly defensive and complex code to handle the limitless possible broken environments users can have.
Defailting to `ip` is the right choice because it's the new, non-deprecated, alternative to `ifconfig`. Defaulting to software which multiple distros are planning to delete is clearly the wrong choice.

I can agree with adding an `osx` check in `has_ip` to claim `osx` never has the tool.
Would such a check be sufficient in your opinion?

Comment (cblecker) [reply build/common.sh]: My concern isn't specific to darwin.. that's just an example. The concern is that we want to ensure the utility we are calling is the one we intend. iproute2 is deprecating net-tools? Great. Then let's make sure the utility we are calling is from iproute2. Even iproute2mac (if installed) on darwin would [match the grep above](https://github.com/brona/iproute2mac/blob/bca8a46108fd7ad55de1a976ff3f7978b116eee9/src/ip.py#L535-L537) and work. But if the "ip" binary doesn't claim to be from iproute2, then it's not the right binary, we can't guarantee that it will respond to `ip address` in the right format, and we should probably fail back to ifconfig.

Or, as I've also suggested, another solution would be refactoring getting the IP address into a function of it's own, and then verifying that you have a valid IP address.

Comment (euank) [reply build/common.sh]: Again, we can't handle every imaginary scenario. It's okay to declare some scenarios as broken dev environments we can ignore.

I think having a bad `ip` binary on linux falls well within that category.
Being cautious/defensive to that degree seems unreasonable to me. It's not based on any real-world linux distros I can think of.

Perhaps a second opinion from an owner of this code, such as @ixdy or @lavalamp would help resolve our apparent disagreement.

Comment (cblecker) [reply build/common.sh]: > It's okay to declare some scenarios as broken dev environments we can ignore.

I'm glad to have other voices contribute to this, but the above statement is something I'd caution against. This would be the equivalent of suggesting that you should `alias ifconfig="ip address"` and be done with it. We want to provide high quality contributor experiences across the board :)

Comment (rphillips) [reply build/common.sh]: @cblecker I just feel like we are getting into the weeds. We do not verify all the userland binaries the scripts shell out to. `which ip` would fail, by default, on a Darwin machine and defer to ifconfig, and the Linux distributions would work flawlessly with both ip and ifconfig installed. 

This style of functionality already exists in this file for the ensure_tar function.

https://github.com/rphillips/kubernetes/blob/4c00b6ed5eaabf938697741d3dda65b52146b9ff/build/common.sh#L295-L314

Comment (cblecker) [reply build/common.sh]: Let's look at this another way then.. Is there a drawback to using `ip -V | grep -o 'iproute2'` instead of `which ip`? I hear that you both feel it's unnecessary, but is there any harms of going that way?

**Comment by rphillips** [comment]
tweaked the has_ip function

File: build/common.sh [hunk 5]
```diff
@@ -317,6 +317,10 @@ function kube::build::has_docker() {
   which docker &> /dev/null
 }
 
+function kube::build::has_ip() {
+  ip -V | grep -o 'iproute2' &> /dev/null
```
Comment (euank) [inline build/common.sh]: nit, `-o` does nothing here. If we're going to do this, we might as well make it clear by using the full flag (`-Version`).

I still feel this adds pointless complexity over `command -v ip` for Linux.
This is a small thing and not that big a deal, but I think it's still worth striving for simplicity and readability.

I won't block on my complaint since it's so minor.

**Comment by rphillips** [comment]
@cblecker can you lgtm again?

File: build/common.sh (hunk 4 above)
Comment (euank) [reply build/common.sh]: Just to explain my position clearly @cblecker. Again, I'm not blocking because getting this change in at all is worth it, but I still think that it would be cleaner code to change the has_ip function to be `! kube::build::is_osx && command -v ip &>/dev/null`

The examples you've given of other commands where the version is checked are not relevant for the `ip` utility. sed and awk are checked because we *know* that there are machines out there without gnu compatible copies of them (such as busybox etc). The reason their version/features are being detected is because we *know* there's a possible problem.

Your argument here is not based on any real possibility. There are no linux machines we *know* of that include a faulty `ip` command. I can use `dnf provides *bin/ip` to tell me that no other fedora package includes the command. If you have evidence to the contrary, please present it, but until you do your argument is not based in reality.

> but is there any harms of going that way?

Writing more complicated code based on what-ifs that have 0 supporting evidence is how you end up with less readable, more complicated, and in general worse code.

The drawback is that this adds unneeded complexity. Someone who comes after will not be sure why this check was added without digging through discussion. It differs from the rest of the codebase (where e.g. [kubenet](https://github.com/kubernetes/kubernetes/blob/v1.8.0/pkg/kubelet/network/kubenet/kubenet_linux.go#L357) already relies on "ip" on linux being "ip").

The code complexity it adds is small, but I don't think we should do things based on "yeah, maybe it's totally pointless, but it's only this little thing". Each bit of pointless code adds up and leads to a worse codebase.

Comment (rphillips) [reply build/common.sh]: This tweak makes sense to me... thoughts @cblecker?

Comment (cblecker) [reply build/common.sh]: @euank If we are concerned about readability, I'd support adding a comment to describe that we are verifying that the ip command is the one provided by iproute2. The argument could be made that even exempting darwin would need a comment as to why we are doing it.

That said, in my opinion, validating that this is the utility we intend to call is not pointless. The point is to ensure that the utility we are calling not only exists, but serves the intended function as we are passing it's output directly into an environment variable.

As I mentioned [here](https://github.com/kubernetes/kubernetes/pull/53500#discussion_r143041596), I think best case is a slight refactor to have an independent function that retrieves and validates the IP address output, before passing it into the docker env variable.

Comment (euank) [reply build/common.sh]: > even exempting darwin would need a comment as to why we are doing it.

Sure. A comment of "ip is a standard Linux utility" should clearly convey why we aren't bothering to check on other distros.

> That said, in my opinion, validating that this is the utility we intend to call is not pointless. The point is to ensure that the utility we are calling not only exists, but serves the intended function

It is a well named standard linux utility. We don't check that `/bin/sh` is posix shell before running shell scripts. We don't check that `docker` is the docker cli (even though there's more evidence it might not be; on debian `docker` can actually be the system tray, [docker](https://packages.debian.org/jessie/x11/docker) which provides a command by the same name).
I've already referenced evidence that this is a standard utility, that there aren't incompatible competing implementations.

On modern Linux machines, running the level of validations that venerable `./configure` scripts run is pointless.
I don't think anything you said properly responds to my comment nor provides evidence contrary to my point.
